### PR TITLE
Cluster frequency coherency check

### PR DESCRIPTION
**Pull request by mdigiorgio**
Usually we work under the assumption that every CPU in a cluster runs at the
same frequency, i.e. frequency changes per-cluster and not per-CPU.

This adds a simple check to prove this assumption true.

Signed-off-by: Michele Di Giorgio <michele.digiorgio@arm.com>

### CODE REVIEW

File: libs/utils/trace.py [hunk 1]
```diff
@@ -377,3 +382,25 @@ def _sanitize_SchedOverutilized(self):
...
+            return
+
+        df = self.df('cpu_frequency')
+        clusters = self.platform['clusters']
+        for c, cpus in clusters.iteritems():
+            cluster_df = df[df.cpu.isin(cpus)]
+            for chunk in self.__chuncker(cluster_df, len(cpus)):
+                f = chunk.iloc[0].frequency
+                if any(chunk.frequency != f):
+                    logging.warn('Cluster Frequency is not coherenti! '\
```
Comment (derkling) [inline libs/utils/trace.py]: nit: s/coherenti/coherent/

File: libs/utils/trace.py [hunk 2]
```diff
@@ -377,3 +382,25 @@ def _sanitize_SchedOverutilized(self):
         df['start'] = df.index
         df['len'] = (df.start - df.start.shift()).fillna(0).shift(-1)
         df.drop('start', axis=1, inplace=True)
+
+    def __chuncker(seq, size):
+        return (seq.iloc[pos:pos + size] for pos in range(0, len(seq), size))
```
Comment (derkling) [inline libs/utils/trace.py]: :heart:

File: libs/utils/trace.py [hunk 3]
```diff
@@ -377,3 +382,25 @@ def _sanitize_SchedOverutilized(self):
         df['start'] = df.index
         df['len'] = (df.start - df.start.shift()).fillna(0).shift(-1)
         df.drop('start', axis=1, inplace=True)
+
+    def __chuncker(seq, size):
```
Comment (JaviMerino) [inline libs/utils/trace.py]: nit: chunker

File: libs/utils/trace.py [hunk 4]
```diff
@@ -70,6 +70,9 @@ def __init__(self, platform, datadir, events,
         # List of events available in the parsed trace
         self.available_events = []
 
+        # Cluster frequency coherency flag
+        self.freq_coherency = True
```
Comment (JaviMerino) [inline libs/utils/trace.py]: What are you going to use this variable for?  Maybe we can leave it off until we have a user for it.

Comment (mdigiorgio) [inline libs/utils/trace.py]: It is used in the frequency residency code because we assume all CPUs in a cluster run at the same frequency.

**Pull request merged**